src/parquet/file/reader-internal.cc

```diff
@@ -273,10 +279,17 @@ void SerializedFile::ParseMetaData() {
 
   std::shared_ptr<PoolBuffer> metadata_buffer =
       AllocateBuffer(properties_.memory_pool(), metadata_len);
-  bytes_read =
-      source_->ReadAt(metadata_start, metadata_len, metadata_buffer->mutable_data());
-  if (bytes_read != metadata_len) {
-    throw ParquetException("Invalid parquet file. Could not read metadata bytes.");
+
+  // Check if the footer_buffer contains the entire metadata
+  if (footer_read_size >= (metadata_len + FOOTER_SIZE)) {
+    memcpy(metadata_buffer->mutable_data(),
+        footer_buffer + (footer_read_size - metadata_len - FOOTER_SIZE), metadata_len);
```

Comment:
This memcpy can be avoided by doing the initial read into allocated memory rather than stack memory. If any buffer slicing needs to be performed, this can be done with zero copy using `arrow::SliceBuffer`. But I don't think this is a big deal either way